### PR TITLE
StructuredDataset without column should be castable to StructuredDataset with columns

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
Upstream sd without column should be castable to sd with columns.
For example, the following task should be successfully compiled by the propeller. 

```python
@task
def to_numpy(ds: StructuredDataset) -> Annotated[StructuredDataset, cols, PARQUET]:
    numpy_array = ds.open(np.ndarray).all()
    return StructuredDataset(dataframe=numpy_array)
```

@task
def t1()

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2074

## Follow-up issue
_NA_

### CODE REVIEW

**Review by wild-endeavor — APPROVED** [review]
+1 from me but we need @kumare3 or @EngHabu

**Review by EngHabu — APPROVED** [review]

**Review by kumare3 — APPROVED** [review]
